ulib/LowStar.Vector.fst

```diff
@@ -0,0 +1,558 @@
+
+val modifies_as_seq:
+  #a:Type -> vec:vector a -> dloc:loc ->
+  h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (live h0 vec /\ 
+		  loc_disjoint (loc_vector vec) dloc /\
+		  modifies dloc h0 h1))
+	(ensures (as_seq h0 vec == as_seq h1 vec))
+	[SMTPat (live h0 vec); 
+	SMTPat (loc_disjoint (loc_vector vec) dloc);
+	SMTPat (modifies dloc h0 h1)]
+let modifies_as_seq #a vec dloc h0 h1 =
+  B.modifies_buffer_elim (Vec?.vs vec) dloc h0 h1
+
+val modifies_as_seq_within:
+  #a:Type -> vec:vector a -> 
+  i:uint32_t -> j:uint32_t{i <= j && j <= size_of vec} ->
+  dloc:loc -> h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (live h0 vec /\ 
+		  loc_disjoint (loc_vector_within vec i j) dloc /\
+		  modifies dloc h0 h1))
+	(ensures (S.slice (as_seq h0 vec) (U32.v i) (U32.v j) == 
+		 S.slice (as_seq h1 vec) (U32.v i) (U32.v j)))
+	[SMTPat (live h0 vec); 
+	SMTPat (loc_disjoint (loc_vector_within vec i j) dloc);
+	SMTPat (modifies dloc h0 h1)]
+let modifies_as_seq_within #a vec i j dloc h0 h1 =
+  B.modifies_buffer_elim (B.gsub (Vec?.vs vec) i (j - i)) dloc h0 h1
+
+/// Construction
+
+val create_empty: 
+  a:Type -> Tot (vec:vector a{size_of vec = 0ul})
+let create_empty a =
+  Vec 0ul 0ul B.null
+
+val create_empty_as_seq_empty:
+  a:Type -> h:HS.mem ->
+  Lemma (S.equal (as_seq h (create_empty a)) S.empty)
+	[SMTPat (as_seq h (create_empty a))]
+let create_empty_as_seq_empty a h = ()
+
+val create_rid:
+  #a:Type -> len:uint32_t{len > 0ul} -> v:a ->
+  rid:HH.rid{HST.is_eternal_region rid} ->
+  HST.ST (vector a)
+	 (requires (fun h0 -> true))
+	 (ensures (fun h0 vec h1 -> 
+	   frameOf vec = rid /\ 
+	   live h1 vec /\ freeable vec /\
+	   modifies loc_none h0 h1 /\
+	   Set.equal (Map.domain (MHS.get_hmap h0))
+                     (Map.domain (MHS.get_hmap h1)) /\
+	   size_of vec = len /\
+	   S.equal (as_seq h1 vec) (S.create (U32.v len) v)))
+let create_rid #a len v rid =
+  Vec len len (B.malloc rid v len)
+
+val create: 
+  #a:Type -> len:uint32_t{len > 0ul} -> v:a -> 
+  HST.ST (vector a)
+	 (requires (fun h0 -> true))
+	 (ensures (fun h0 vec h1 -> 
+	   frameOf vec = HH.root /\
+	   live h1 vec /\ freeable vec /\
+	   modifies loc_none h0 h1 /\
+	   Set.equal (Map.domain (MHS.get_hmap h0))
+                     (Map.domain (MHS.get_hmap h1)) /\
+	   size_of vec = len /\
+	   S.equal (as_seq h1 vec) (S.create (U32.v len) v)))
+let create #a len v =
+  create_rid len v HH.root
+
+val create_reserve:
+  #a:Type -> len:uint32_t{len > 0ul} -> ia:a ->
+  rid:HH.rid{HST.is_eternal_region rid} ->
+  HST.ST (vector a)
+	 (requires (fun h0 -> true))
+	 (ensures (fun h0 vec h1 -> 
+	   frameOf vec = rid /\ live h1 vec /\ freeable vec /\
+	   modifies loc_none h0 h1 /\
+	   Set.equal (Map.domain (MHS.get_hmap h0))
+                     (Map.domain (MHS.get_hmap h1)) /\
+	   size_of vec = 0ul /\
+	   S.equal (as_seq h1 vec) S.empty))
+let create_reserve #a len ia rid =
+  Vec 0ul len (B.malloc rid ia len)
+
+val create_by_buffer:
```

Comment:
Tahina recommends naming these `create*` based on the fact that `alloc`/`free` oftentimes go hand-in-hand

Reply:
(just like in the `Buffer` library)